### PR TITLE
Generate crd by controller-gen tools

### LINKED ISSUE #4032: apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
## Bug Report
apiextensions.k8s.io/v1beta1 is deprecated, is there any plan to replace with v1?

**What version of Kubernetes are you using?**
1.20

**What did you do?**
kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/v1.1.12/manifests/crd.yaml

**What did you expect to see?**
no warning messages

**What did you see instead?**
image

### PR DESCRIPTION
**Pull request by liubog2008**
### What problem does this PR solve?

Fix https://github.com/pingcap/tidb-operator/issues/2753 Fix #4032 

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [x] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### TODO

- [x] add crd generation script
- [x] add generated crd
- [x] clean useless code

### CODE REVIEW

**Comment by liubog2008** [comment]
@DanielZhangQD PTAL

File: manifests/crd/v1/pingcap.com_dataresources.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+
+---
```
Comment (DanielZhangQD) [inline manifests/crd/v1/pingcap.com_dataresources.yaml:2]: No need to generate crd for this.

File: manifests/crd/v1beta1/pingcap.com_dataresources.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
+
+---
```
Comment (DanielZhangQD) [inline manifests/crd/v1beta1/pingcap.com_dataresources.yaml:2]: ditto.

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 3]
```diff
@@ -909,7 +913,7 @@ type ServiceSpec struct {
 // +k8s:openapi-gen=true
 type TiDBServiceSpec struct {
 	// +k8s:openapi-gen=false
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:915]: Should this be changed to `true`?

**Comment by DanielZhangQD** [comment]
@liubog2008 The `hack/verify-crd-groups.sh` needs to be updated to make sure that the CRDs are generated correctly if any changes to the `types`.

File: manifests/crd/v1/pingcap.com_dataresources.yaml (hunk 1 above)
Comment (liubog2008) [reply manifests/crd/v1/pingcap.com_dataresources.yaml:2]: DataResource is not used?

Comment (DanielZhangQD) [reply manifests/crd/v1/pingcap.com_dataresources.yaml:2]: No.

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 3 above)
Comment (liubog2008) [reply pkg/apis/pingcap/v1alpha1/types.go:915]: can be changed in another PR? This does not relate to crd.

Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/types.go:915]: Yes, it's not related to CRD generation, however, with `false`, the following fields will not be shown in the final CRD files, what about we just fix this in this PR?

File: manifests/crd/v1/pingcap.com_dataresources.yaml (hunk 1 above)
Comment (liubog2008) [reply manifests/crd/v1/pingcap.com_dataresources.yaml:2]: It seems that this crd exists in `manifests/data-resource-crd.yaml`

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 3 above)
Comment (liubog2008) [reply pkg/apis/pingcap/v1alpha1/types.go:915]: All fields in `ServiceSpec` have been generated in current CRD.

This comment only affect openapi spec generation.

**Review by july2993 — COMMENTED** [review]
after this pr, we can support v1.22 and older version k8s both?

**Comment by liubog2008** [comment]
@july2993 Yes

File: Makefile [hunk 4]
```diff
@@ -160,16 +160,19 @@ test:
...
-check: $(addprefix check-,$(ALL_CHECKS)) lint tidy 
+check: $(addprefix check-,$(ALL_CHECKS)) lint tidy
 
 check-%:
 	./hack/verify-$*.sh
 
 lint:
 	./hack/verify-lint.sh
 
+crd:
```
Comment (DanielZhangQD) [inline Makefile:173]: Add a target for `api-references`?

Comment (liubog2008) [reply Makefile:173]: another pr ?

**Comment by liubog2008** [comment]
@DanielZhangQD PTAL

**Comment by liubog2008** [comment]
@DanielZhangQD Tests failed because image registry is unauthorized.

**Comment by liubog2008** [comment]
@DanielZhangQD All tests are passed

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 5]
```diff
@@ -100,17 +101,35 @@ const (
...
 	metav1.TypeMeta `json:",inline"`
 	// +k8s:openapi-gen=false
-	metav1.ObjectMeta `json:"metadata"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
 
 	// Spec defines the behavior of a tidb cluster
-	Spec TidbClusterSpec `json:"spec"`
+	// +optional
+	Spec TidbClusterSpec `json:"spec,omitempty"`
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/types.go:127]: why do not mark these fields as required anymore?

File: hack/crdgen/go.mod [hunk 6]
```diff
@@ -0,0 +1,5 @@
+module github.com/pingcap/tidb-operator/hack/crdgen
+
+go 1.15
```
[the remark this answers is not in the document]
Comment (liubog2008) [reply hack/crdgen/go.mod:3]: Same version is ok, I'll change it

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 5 above)
Comment (liubog2008) [reply pkg/apis/pingcap/v1alpha1/types.go:127]: emmm, It seems all `spec` fields of kubernetes resource are set as optional. I can't find any documents explain why.

Comment (csuzhangxc) [reply pkg/apis/pingcap/v1alpha1/types.go:127]: but I think some fields like `spec` should be `Required`. now, `Required: []string{"spec"},` has been removed in generated open API code.

**Comment by DanielZhangQD** [comment]
Close via https://github.com/pingcap/tidb-operator/pull/4151.

**Pull request closed without merging**